Doc/library/sqlite3.rst

```diff
@@ -501,11 +501,7 @@ Module constants
 
    .. note::
 
-      The :mod:`!sqlite3` module supports ``qmark``, ``numeric``,
-      and ``named`` DB-API parameter styles,
-      because that is what the underlying SQLite library supports.
-      However, the DB-API does not allow multiple values for
-      the ``paramstyle`` attribute.
```

Comment:
I find these lines very confusing; IMO we're better off without them.

cc. @CAM-Gerlach

Reply:
agree; you might want to clarify though that the use of "qmark" or "named" is detected automatically on a per-statement basis (provided my understanding of that is correct).

Reply:
> I find these lines very confusing; IMO we're better off without them.

Agreed; as a relative n00b I was particularly confused, but assumed this was just subject matter details that flew right over my head.

Reply:
> agree; you might want to clarify though that the use of "qmark" or "named" is detected automatically on a per-statement basis (provided my understanding of that is correct).

Yes, we should definitely clarify how parameters are interpreted and mapped to the placeholders.

`sqlite3` does not check if you use the "qmark" or "named" style (or any other style FWIW[^1]); it only looks at the type of the params supplied:

1. If a dict or dict subclass is supplied, the named style is assumed and you'll get an error if a named parameter is not provided by the supplied dict.

2. If an exact tuple, an exact list, or a sequence (that is not a dict or dict subclass) is supplied, the qmark style[^2] is assumed. This means that `sqlite3` iterates over the params and blindly assigns placeholder 1[^3] the first item in the supplied sequence, and so on. This also happens if you use named placeholders and supply, for example, a list. Try it and be surprised. Now, that bug may be too old to be fixed; there's bound to be some code out there that depends on this exact bug. We might be able to introduce a warning and then change the behaviour after a few release cycles, but such a breaking change/bugfix will need a broader discussion.

[^1]: try for example `cx.execute("select ?2, ?1", ['first', 'second'])`; the SQLite numeric style, which is not PEP-249-compatible, is accepted and correctly applied
[^2]: called nameless in SQLite speak
[^3]: SQLite placeholders use one-based indices

Reply:
> 1. If a dict or dict subclass is supplied, the named style is assumed and you'll get an error if a named parameter is not provided by the supplied dict.

☝🏻 Note that the checks are for `dict` or `dict` subclass; it is not a mapping check.

Reply:
I've created a follow-up issue for this: #100668